### PR TITLE
Refactor datasets API

### PR DESCRIPTION
**Pull request by davelopez**
Preliminary refactorings before migrating this API controller to FastAPI.
- [x] Move controller logic to the service layer
- [x] Add type annotations
- [x] Create pydantic models for payloads and responses

## How to test the changes?
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Comment by davelopez** [comment]
I think this is ready for review.
I will follow up with another PR for the actual creation of FastAPI routes with the documentation, etc.

**Comment by davelopez** [comment]
Any ideas of what may be wrong with [test_create_anonymous_fails](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy_test/api/test_histories.py#L194)?
It failed again after a rerun. When run locally it passes (most of the time) but sometimes it fails with this `Connection reset by peer`. Not sure if it's related to how anonymous access is handled in testing, but I think I've seen from time to time also failing [test_nonadmin](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy_test/api/test_libraries.py#L49) for similar reasons :thinking:

**Comment by mvdbeek** [comment]
I think that only fails with the legacy tests, also in other PRs:
```
E               requests.exceptions.ChunkedEncodingError: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
```

File: lib/galaxy/webapps/galaxy/api/common.py [hunk 1]
```diff
@@ -42,3 +43,16 @@ def query_serialization_params(
     default_view: Optional[str] = SerializationDefaultViewQueryParam,
 ) -> SerializationParams:
     return parse_serialization_params(view=view, keys=keys, default_view=default_view)
+
+
+def get_update_permission_payload(payload: Dict[str, Any]) -> UpdateDatasetPermissionsPayload:
+    """Coverts the generic payload dictionary into a UpdateDatasetPermissionsPayload model with custom parsing.
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/common.py:49]:
```suggestion
    """Converts the generic payload dictionary into a UpdateDatasetPermissionsPayload model with custom parsing.
```

**Review by mvdbeek — APPROVED** [review]
There's a minor typo, but we can followup in #12924

**Comment by mvdbeek** [comment]
Very nice, I like how clean the API only layer is.

**Comment by davelopez** [comment]
Thank you Marius!

**Pull request merged**